json2: C error on encoding optional anon struct field

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 3271c72, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 89d1aac.3271c72
|:-------------------|:-------------------
|OS                  |linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |0.64GB/15.35GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2025-07-02 13:14:42
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /tmp
|                    |
|Git version         |git version 2.50.0
|V git status        |weekly.2025.16-385-g3271c728
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.1.1 20250425
|gcc version         |gcc (GCC) 15.1.1 20250425
|clang version       |clang version 20.1.6
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.41

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  anon_struct_json2.v && anon_struct_json2`
```v
import x.json2

struct Disk {
	dev  string
	size ?struct {
		value u64
	}
}

fn main() {
	disk := Disk{}
	disk_str := json2.encode[Disk](disk)
	dump(disk_str)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/anon_struct_json2.01JZ5N9Y5NBT75FH8QRJRM2RHG.tmp.c:11152: error: 'U' undeclared
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

`{"dev":""}`

---

`json` works as expected (but fails with `-skip-unused` in short program):
```v
import json

struct Disk {
	dev  string
	size ?struct {
		value u64
	}
}

fn main() {
	disk := Disk{}
	disk_str := json.encode(disk)
	dump(disk_str)
}
```
```
/tmp $ v -no-skip-unused run anon_struct_json.v
[anon_struct_json.v:13] disk_str: {"dev":""}
```


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2767,7 +2767,8 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.expr_with_cast(arg.expr, arg_typ, expected_type)
 		g.write('.data')
 		return
-	} else if arg.expr is ast.Ident && arg_sym.info is ast.Struct && arg_sym.info.is_anon {
+	} else if arg.expr is ast.Ident && arg_sym.info is ast.Struct && arg_sym.info.is_anon
+		&& !expected_type.has_flag(.generic) {
 		// make anon struct struct compatible with another anon struct declaration
 		g.write('*(${g.cc_type(expected_type, false)}*)&')
 	}

--- a/vlib/x/json2/tests/encode_anon_struct_test.v
+++ b/vlib/x/json2/tests/encode_anon_struct_test.v
@@ -1,0 +1,24 @@
+import x.json2
+
+struct Disk {
+	dev  string
+	size ?struct {
+		value u64
+	}
+}
+
+fn test_main() {
+	disk := Disk{
+		size: struct {
+			value: 123
+		}
+	}
+	disk_str := json2.encode[Disk](disk)
+	assert disk_str == '{"dev":"","size":{"value":123}}'
+}
+
+fn test_none() {
+	disk := Disk{}
+	disk_str := json2.encode[Disk](disk)
+	assert disk_str == '{"dev":""}'
+}

